File: frontend/src/models/action/action.ts
import { Character } from '../card/card';

export enum ActionType {
  INCOME = 'income',
  FOREIGN_AID = 'foreign_aid',
  COUP = 'coup',
  TAX = 'tax',
  ASSASSINATE = 'assassinate',
  STEAL = 'steal',
  EXCHANGE = 'exchange',
}

export interface ActionRule {
  type: ActionType;
  label: string;
  description: string;
  cost: number;
  characterRequired: Character | null;
  isChallengeble: boolean;
  blockedBy: Character[];
  requiresTarget: boolean;
}

export interface ActionPresentation {
  accent: string;
  tint: string;
  icon: 'income' | 'aid' | 'coup' | 'tax' | 'assassinate' | 'steal' | 'exchange';
}

/**
 * Single source of truth for all action rules on the frontend.
 * Mirrors backend ACTION_RULES.
 */
export const ACTION_RULES: Record<ActionType, ActionRule> = {
  [ActionType.INCOME]: {
    type: ActionType.INCOME,
    label: 'Income',
    description: 'Take 1 coin from the treasury.',
    cost: 0,
    characterRequired: null,
    isChallengeble: false,
    blockedBy: [],
    requiresTarget: false,
  },
  [ActionType.FOREIGN_AID]: {
    type: ActionType.FOREIGN_AID,
    label: 'Foreign Aid',
    description: 'Take 2 coins from the treasury.',
    cost: 0,
    characterRequired: null,
    isChallengeble: false,
    blockedBy: [Character.DUKE],
    requiresTarget: false,
  },
  [ActionType.COUP]: {
    type: ActionType.COUP,
    label: 'Coup',
    description: 'Pay 7 coins to force a player to lose an influence.',
    cost: 7,
    characterRequired: null,
    isChallengeble: false,
    blockedBy: [],
    requiresTarget: true,
  },
  [ActionType.TAX]: {
    type: ActionType.TAX,
    label: 'Tax',
    description: 'Bluff Duke — take 3 coins from the treasury.',
    cost: 0,
    characterRequired: Character.DUKE,
    isChallengeble: true,
    blockedBy: [],
    requiresTarget: false,
  },
  [ActionType.ASSASSINATE]: {
    type: ActionType.ASSASSINATE,
    label: 'Assassinate',
    description: 'Bluff Assassin — pay 3 coins to kill an influence.',
    cost: 3,
    characterRequired: Character.ASSASSIN,
    isChallengeble: true,
    blockedBy: [Character.CONTESSA],
    requiresTarget: true,
  },
  [ActionType.STEAL]: {
    type: ActionType.STEAL,
    label: 'Steal',
    description: 'Bluff Captain — take 2 coins from another player.',
    cost: 0,
    characterRequired: Character.CAPTAIN,
    isChallengeble: true,
    blockedBy: [Character.CAPTAIN, Character.AMBASSADOR],
    requiresTarget: true,
  },
  [ActionType.EXCHANGE]: {
    type: ActionType.EXCHANGE,
    label: 'Exchange',
    description: 'Bluff Ambassador — swap cards with the Court Deck.',
    cost: 0,
    characterRequired: Character.AMBASSADOR,
    isChallengeble: true,
    blockedBy: [],
    requiresTarget: false,
  },
};

export const ACTION_PRESENTATIONS: Record<ActionType, ActionPresentation> = {
  [ActionType.INCOME]: {
    accent: '#D8B24A',
    tint: 'rgba(216, 178, 74, 0.14)',
    icon: 'income',
  },
  [ActionType.FOREIGN_AID]: {
    accent: '#7DB8D6',
    tint: 'rgba(125, 184, 214, 0.15)',
    icon: 'aid',
  },
  [ActionType.COUP]: {
    accent: '#E77F67',
    tint: 'rgba(231, 127, 103, 0.15)',
    icon: 'coup',
  },
  [ActionType.TAX]: {
    accent: '#C79AE8',
    tint: 'rgba(199, 154, 232, 0.16)',
    icon: 'tax',
  },
  [ActionType.ASSASSINATE]: {
    accent: '#E96A74',
    tint: 'rgba(233, 106, 116, 0.15)',
    icon: 'assassinate',
  },
  [ActionType.STEAL]: {
    accent: '#65B5E8',
    tint: 'rgba(101, 181, 232, 0.15)',
    icon: 'steal',
  },
  [ActionType.EXCHANGE]: {
    accent: '#6FC59B',
    tint: 'rgba(111, 197, 155, 0.15)',
    icon: 'exchange',
  },
};

File: backend/app/models/action.py
"""Action models — single source of truth for game actions and their rules."""

from __future__ import annotations

from enum import Enum

from pydantic import BaseModel, ConfigDict

from app.models.card import Character


class ActionType(str, Enum):
    """All possible actions a player can take."""

    INCOME = "income"
    FOREIGN_AID = "foreign_aid"
    COUP = "coup"
    TAX = "tax"
    ASSASSINATE = "assassinate"
    STEAL = "steal"
    EXCHANGE = "exchange"


class ActionRule(BaseModel):
    """Defines the rules for a single action type."""

    model_config = ConfigDict(strict=True)

    action_type: ActionType
    cost: int = 0
    character_required: Character | None = None
    is_challengeable: bool = False
    blocked_by: list[Character] = []
    requires_target: bool = False


# ── Single source of truth: action rules lookup ──────────────────────────────

ACTION_RULES: dict[ActionType, ActionRule] = {
    ActionType.INCOME: ActionRule(
        action_type=ActionType.INCOME,
        cost=0,
        character_required=None,
        is_challengeable=False,
        blocked_by=[],
        requires_target=False,
    ),
    ActionType.FOREIGN_AID: ActionRule(
        action_type=ActionType.FOREIGN_AID,
        cost=0,
        character_required=None,
        is_challengeable=False,
        blocked_by=[Character.DUKE],
        requires_target=False,
    ),
    ActionType.COUP: ActionRule(
        action_type=ActionType.COUP,
        cost=7,
        character_required=None,
        is_challengeable=False,
        blocked_by=[],
        requires_target=True,
    ),
    ActionType.TAX: ActionRule(
        action_type=ActionType.TAX,
        cost=0,
        character_required=Character.DUKE,
        is_challengeable=True,
        blocked_by=[],
        requires_target=False,
    ),
    ActionType.ASSASSINATE: ActionRule(
        action_type=ActionType.ASSASSINATE,
        cost=3,
        character_required=Character.ASSASSIN,
        is_challengeable=True,
        blocked_by=[Character.CONTESSA],
        requires_target=True,
    ),
    ActionType.STEAL: ActionRule(
        action_type=ActionType.STEAL,
        cost=0,
        character_required=Character.CAPTAIN,
        is_challengeable=True,
        blocked_by=[Character.CAPTAIN, Character.AMBASSADOR],
        requires_target=True,
    ),
    ActionType.EXCHANGE: ActionRule(
        action_type=ActionType.EXCHANGE,
        cost=0,
        character_required=Character.AMBASSADOR,
        is_challengeable=True,
        blocked_by=[],
        requires_target=False,
    ),
}


class PlayerAction(BaseModel):
    """An action submitted by a player."""

    model_config = ConfigDict(strict=True)

    action_type: ActionType
    player_id: str
    target_id: str | None = None


class ActionResult(BaseModel):
    """The result after an action resolves."""

    model_config = ConfigDict(strict=True)

    success: bool
    action: PlayerAction
    coins_changed: dict[str, int] = {}  # player_id -> delta
    influences_lost: dict[str, int] = {}  # player_id -> card_index
    message: str = ""
